### PR TITLE
Add interfaces

### PR DESCRIPTION
**Pull request by bram-pkg**

### CODE REVIEW

File: src/Extension/CompanyExtension.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+     * @example 'Acme Ltd'
+     */
+    public function company(): string;
+
+    /**
+     * @example 'Ltd'
+     */
+    public function companySuffix(): string;
+
+    public function jobTitle(): string;
```
Comment (pimjansen) [inline src/Extension/CompanyExtension.php:20]: Lets add an example here too for reference

File: src/Extension/DateTimeExtension.php [hunk 2]
```diff
@@ -0,0 +1,240 @@
+<?php
+
+namespace Faker\Core\Extension;
+
+/**
+ * FakerPHP extension for Date-related randomization.
+ *
+ * Functions accepting a date string use the `strtotime()` function internally.
+ *
+ * @experimental
```
Comment (pimjansen) [inline src/Extension/DateTimeExtension.php:10]: cleanup phpdoc

File: src/Extension/DateTimeExtension.php [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+     * @param \DateTime|int|string $until    maximum timestamp, defaults to "now"
+     * @param string|null          $timezone zone timezone for generated date, fallback to `DateTime::$defaultTimezone` and `date_default_timezone_get()`.
+     *
+     * @see \DateTimeZone
+     * @see http://php.net/manual/en/timezones.php
+     * @see http://php.net/manual/en/function.date-default-timezone-get.php
+     *
+     * @example DateTime('2005-08-16 20:39:21')
+     */
+    public function dateTime($until = 'now', string $timezone = null): \DateTime;
```
Comment (pimjansen) [inline src/Extension/DateTimeExtension.php:28]:
```suggestion
    public function dateTime(string $until = 'now', string $timezone = null): \DateTime;
```

File: src/Extension/DateTimeExtension.php [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+     *
+     * @param \DateTime|int|string $until    maximum timestamp, defaults to "now"
+     * @param string|null          $timezone zone timezone for generated date, fallback to `DateTime::$defaultTimezone` and `date_default_timezone_get()`.
+     *
+     * @example DateTime('1265-03-22 21:15:52')
+     *
+     * @see http://php.net/manual/en/timezones.php
+     * @see http://php.net/manual/en/function.date-default-timezone-get.php
+     */
+    public function dateTimeAD($until = 'now', string $timezone = null): \DateTime;
```
Comment (pimjansen) [inline src/Extension/DateTimeExtension.php:41]:
```suggestion
    public function dateTimeAD(string $until = 'now', string $timezone = null): \DateTime;
```

File: src/Extension/DateTimeExtension.php [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+     *
+     * @param \DateTime|string     $from     defaults to 30 years ago
+     * @param \DateTime|int|string $until    maximum timestamp, defaults to "now"
+     * @param string|null          $timezone zone timezone for generated date, fallback to `DateTime::$defaultTimezone` and `date_default_timezone_get()`.
+     *
+     * @see \DateTimeZone
+     * @see http://php.net/manual/en/timezones.php
+     * @see http://php.net/manual/en/function.date-default-timezone-get.php
+     */
+    public function dateTimeBetween($from = '-30 years', $until = 'now', string $timezone = null): \DateTime;
```
Comment (pimjansen) [inline src/Extension/DateTimeExtension.php:55]:
```suggestion
    public function dateTimeBetween(string $from = '-30 years', string $until = 'now', string $timezone = null): \DateTime;
```
I do wonder though if it isnt better to just add a DateTimeInterface in the from and untill

File: src/Extension/GeneratorAwareExtensionTrait.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * A helper trait to be used with GeneratorAwareExtension.
+ */
+trait GeneratorAwareExtensionTrait
+{
+    /**
+     * @var Generator|null
+     */
+    private $generator;
```
Comment (pimjansen) [inline src/Extension/GeneratorAwareExtensionTrait.php:18]:
```suggestion
    private ?Generator $generator;
```

File: src/Extension/Helper.php [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+ * A class with some methods that may make building extensions easier.
+ *
+ * @experimental This class is experimental and does not fall under our BC promise
+ */
+final class Helper
+{
+    /**
+     * Returns a random element from a passed array.
+     */
+    public static function randomElement(array $array)
```
Comment (pimjansen) [inline src/Extension/Helper.php]:
```suggestion
    public static function randomElement(array $array): mixed
```

File: src/Extension/UuidExtension.php [hunk 8]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace Faker\Core\Extension;
+
+/**
+ * @experimental This interface is experimental and does not fall under our BC promise
+ */
+interface UuidExtension extends Extension
```
Comment (pimjansen) [inline src/Extension/UuidExtension.php]: As discussed, lets not include this. The UUID generation is a bit flakey anyhow and only supports v3 where there are multiple versions. Lets suggest people to use an actual UUID lib to generate those

File: src/Generator.php [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+     *
+     * @param bool $reset If set to true, resets the list of existing values
+     * @param int $maxRetries Maximum number of retries to find a unique value,
+     *                         After which an OverflowException is thrown.
+     *
+     * @return self A proxy class returning only non-existing values
+     * @throws \OverflowException When no unique value can be found by iterating $maxRetries times
+     *
+     */
+    public function unique($reset = false, $maxRetries = 10000)
```
Comment (pimjansen) [inline src/Generator.php]:
```suggestion
    public function unique(bool $reset = false, int $maxRetries = 10000): UniqueGenerator
```

File: src/Generator.php [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+     * @return self A proxy class returning only valid values
+     * @throws \OverflowException When no valid value can be found by iterating $maxRetries times
+     *
+     */
+    public function valid(?\Closure $validator = null, int $maxRetries = 10000)
+    {
+        return new ValidGenerator($this, $validator, $maxRetries);
+    }
+
+    public function seed($seed = null): void
```
Comment (pimjansen) [inline src/Generator.php]:
```suggestion
    public function seed(int $seed = null): void
```

File: src/Generator.php [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+    {
+        if ($seed === null) {
+            mt_srand();
+        } else {
+            mt_srand((int)$seed, MT_RAND_PHP);
```
Comment (pimjansen) [inline src/Generator.php]:
```suggestion
            mt_srand($seed, MT_RAND_PHP);
```

File: src/Generator.php [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    /**
+     * Replaces tokens ('{{ tokenName }}') with the result from the token method call
+     *
+     * @param string $string String that needs to bet parsed
+     *
+     * @return string
+     */
+    public function parse($string)
```
Comment (pimjansen) [inline src/Generator.php]:
```suggestion
    public function parse(string $string): string
```

File: src/Implementation/Blood.php [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * @experimental This class is experimental and does not fall under our BC promise
+ */
+final class Blood implements BloodExtension
+{
+    /**
+     * @var string[]
+     */
+    private $bloodTypes = ['A', 'AB', 'B', 'O'];
```
Comment (pimjansen) [inline src/Implementation/Blood.php:18]:
```suggestion
    private array $bloodTypes = ['A', 'AB', 'B', 'O'];
```

File: src/Implementation/Blood.php [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * @var string[]
+     */
+    private $bloodRhFactors = ['+', '-'];
```
Comment (pimjansen) [inline src/Implementation/Blood.php:23]:
```suggestion
    private array $bloodRhFactors = ['+', '-'];
```

File: src/Extension/DateTimeExtension.php (hunk 3 above)
Comment (bram-pkg) [reply src/Extension/DateTimeExtension.php:28]: It can be int, string, and some others too.

**Review by pimjansen — APPROVED** [review]
Ok for now, did not do a full deepdive but lets adjust later where needed

**Pull request merged**